dvc/utils/http.py

```diff
@@ -0,0 +1,100 @@
+from contextlib import contextmanager
+import io
+
+import requests
+
+from dvc.utils.compat import is_py2
+
+
+@contextmanager
+def open_url(url, mode="r", encoding=None):
+    """Opens an url as a readable stream.
+
+    Resumes on connection error.
+    Url could be a string or a callable returning a string.
+    """
+    assert mode in {"r", "rt", "rb"}
+
+    with iter_url(url) as (response, it):
+        bytes_stream = IterStream(it)
+
+        if mode == "rb":
+            yield bytes_stream
+        else:
+            encoding = encoding or response.encoding
+            yield io.TextIOWrapper(bytes_stream, encoding=encoding)
+
+
+@contextmanager
+def iter_url(url, chunk_size=io.DEFAULT_BUFFER_SIZE):
+    """Iterate over chunks requested from url."""
+
+    def request(headers=None):
+        the_url = url() if callable(url) else url
+        response = requests.get(the_url, stream=True, headers=headers)
+        response.raise_for_status()
+        return response
+
+    def gen(response):
+        try:
+            pos = 0
+            while True:
+                try:
+                    for chunk in response.iter_content(chunk_size):
+                        pos += len(chunk)
+                        yield chunk
+                    break
+                except requests.ConnectionError:
+                    response.close()
+                    if response.headers.get("Accept-Ranges") != "bytes":
+                        raise
+
+                    # Reopen request from where we stopped
+                    headers = {"Range": "bytes={}-".format(pos)}
+                    response = request(headers)
+        finally:
+            response.close()
+
+    response = request()
+    it = gen(response)
+    try:
+        yield response, it
+    finally:
+        # Ensure connection is closed
+        it.close()
+
+
+class IterStream(io.RawIOBase):
+    """Wraps an iterator yielding bytes as a file object"""
+
+    def __init__(self, iterator):
+        self.iterator = iterator
+        self.leftover = None
+
+    def readable(self):
+        return True
+
+    def readinto(self, b):
+        try:
+            b_len = len(b)  # We're supposed to return at most this much
+            chunk = self.leftover or next(self.iterator)
+            output, self.leftover = chunk[:b_len], chunk[b_len:]
+            b[: len(output)] = output
+            return len(output)
+        except StopIteration:
+            return 0  # indicate EOF
+
+    if is_py2:
+
+        def read1(self, n=-1):
+            try:
+                chunk = self.leftover or next(self.iterator)
+            except StopIteration:
+                return b""
+
+            if n > 0:
+                output, self.leftover = chunk[:n], chunk[n:]
+                return output
+            else:
```

Comment:
As DS noted, else is not needed.

Reply:
I think it's more obvious this way.

Reply:
It is not and we don't do such things in other places. Please adjust.